app/models/report.rb

```diff
@@ -4,14 +4,33 @@ class Report < ApplicationRecord
   belongs_to :user
   has_many :comments, as: :commentable, dependent: :destroy
 
+  has_many :report_links, dependent: :destroy
+  has_many :to_reports, through: :report_links, source: :to_report
+  has_many :my_report_links, class_name: 'ReportLink', foreign_key: 'to_report_id', dependent: :destroy, inverse_of: :to_report
+  has_many :reports, through: :my_report_links, source: :report
+
   validates :title, presence: true
   validates :content, presence: true
 
+  def link(to_report)
+    return if self == to_report
+
+    report_links.find_or_create_by(to_report_id: to_report.id)
```

Comment:
```suggestion
    report_links.find_or_create_by!(to_report_id: to_report.id)
```

のように`!`を付けて保存に失敗した場合は気付きやすくした方がいいかもしれません